New Resource: `azurerm_monitor_scheduled_query_rules_alert` & `azurerm_monitor_scheduled_query_rules_log`

mcdafydd opened the pull request:
Fixes #3951

This feature implements the [ScheduledQueryRules API](https://docs.microsoft.com/en-us/rest/api/monitor/scheduledqueryrules).  The new resources should allow users to automate creation of monitoring and alerting rules alongside infrastructure.

mcdafydd commented:
I think I'm all good now and ready for review.

Thanks!

anvarich commented:
@tombuildsstuff @katbyte would be great to have it 1.39.0 release ;)

anvarich commented:
NY present maybe in this time 🙏

SIGAN commented:
Please review this pull request - many people waiting for this change!

anvarich commented:
@katbyte as i understood from tagging it will be merged in current 1.42 release cycle?

katbyte commented on azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go:
```diff
@@ -0,0 +1,658 @@
+					"LogToMetric",
+				}, false),
+			},
+			"authorized_resources": {
```
Are these ids? if so 
```suggestion
			"authorized_resource_ids": {
```

might be more clear?

mcdafydd replied:
Done!

katbyte commented on azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go:
```diff
@@ -0,0 +1,658 @@
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
```
Can we validate these?

katbyte commented on azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go:
```diff
@@ -0,0 +1,658 @@
+					Type: schema.TypeString,
+				},
+			},
+			"azns_action": {
```
azns sounds fairly opaque, is there a better name we could use here?

mcdafydd replied:
I renamed this from azns_action to simply `action`

katbyte commented on azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go:
```diff
@@ -0,0 +1,658 @@
+							Required: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
```
Could we validate this?

katbyte commented on azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go:
```diff
@@ -0,0 +1,658 @@
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"action_group": {
```
is this the name or id?

katbyte commented on azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go:
```diff
@@ -0,0 +1,658 @@
+									"name": {
+										Type:     schema.TypeString,
+										Required: true,
+									},
```
COuld we validate this with at least no empty stings?

katbyte commented on azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go:
```diff
@@ -0,0 +1,658 @@
+									},
+									"operator": {
+										Type:     schema.TypeString,
+										Required: true,
```
If this has only one value, we could make this optional with a default?

katbyte commented on azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go:
```diff
@@ -0,0 +1,658 @@
+										Required: true,
+										Elem: &schema.Schema{
+											Type: schema.TypeString,
+										},
```
COuld we validate this with at least no empty stings?

katbyte commented on azurerm/internal/services/monitor/resource_arm_monitor_scheduled_query_rules.go:
```diff
@@ -0,0 +1,658 @@
+						"metric_name": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
```
COuld we validate this with at least no empty stings?

katbyte commented on azurerm/internal/services/monitor/tests/data_source_monitor_scheduled_query_rules_test.go:
```diff
@@ -0,0 +1,243 @@
+func testAccDataSourceAzureRMMonitorScheduledQueryRules_logToMetricActionConfig(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-%d"
```
could we use
```suggestion
  name     = "acctestRG-monitor-%d"
```

for allr esource group names?

mcdafydd commented:
I believe all the comments are addressed.  I requested a review but that didn't seem to remove the "waiting-response" label.

davidtom commented on website/docs/r/monitor_scheduled_query_rules_alert.html.markdown:
```diff
@@ -0,0 +1,153 @@
+  severity       = 1
+  time_window    = 30
+  trigger {
+    threshold_operator = "GreaterThan"
```
```suggestion
    operator = "GreaterThan"
```

davidtom replied:
When using this resource I get an error that `threshold_operator` is not supported and that I'm missing the `operator` argument

mcdafydd replied:
Thank you!  I pushed an update.

mcdafydd commented:
> Hey @mcdafydd, I've got a doozy of a review for you but I appreciate the effort so far. My main concern is around maintainability and I've laid out the reasoning for that in a comment down below.

@mbfrahry, I just want to say thank you very much for the thoughtful review and the time to share your insights.  I'll start addressing these issues asap and try to push an update later tonight or tomorrow.  All your comments make sense to me.

I'm surprised by the crashing test.  I may need to update from master.

mcdafydd commented:
I think we're getting closer.  I addressed a lot of the concerns you brought up, but still have a bit more to do on validations, helper functions, and tests.

For validations, I followed comments that appeared in the Azure portal and anything I could find in the documentation or when testing the APIs.

I'm thinking about creating a third resource called `azurerm_monitor_scheduled_query_rules_alert_metric`.  There seem to be three types of things you can do with the ScheduledQueryRules API today:

1. LogToMetricAction - near-realtime; stateful; limited set of supported tables (ie: Perf, Event, Heartbeat, Update) - creates an Azure Metric from a Log Analytics resource that can be alerted upon using a Metric Alert rule
2. AlertingAction with result count - stateless; runs action if the number of query results exceeds threshold
3. AlertingAction with metric trigger - stateless; requires time series aggregation and evaluates query results by metric column name and operator; runs action if the consecutive or total number of metric results exceeds threshold

The Azure portal gives this error when trying to create a metric trigger if the query isn't correct: `Search Query should contain 'AggregatedValue' and 'bin(TimeGenerated, [roundTo])' for Metric alert type`.  I was thinking that a new resource would be better here and allow us to validate the search query in the schema rather than in the CreateUpdate() function, but I'm open to suggestions.

mbfrahry commented:
Glad to hear it @mcdafydd! 

I appreciate you taking the time to go through the validations. We try to keep it consistent with the UI and API but when it's too difficult to find we can just go with sane validation like no empty strings or integers greater than 0. The more precise the better though.

As for the third resource, we're dipping into even more complicated territory. We have too many potential routes to go down and no real best option. I'm hesitant to create a whole new resource if the only difference between the two is what the query can have in it. Are there more differences than just that?  It'd be hard to police from an import perspective as someone who doesn't understand this to the fullest could import incorrectly and then be forced to untangle how their resources should be saved in state. I think for this case just having an extra example in the documentation that shows how to use the trigger with the correct query will suffice. We could add additional validation around triggers and queries in CreateUpdate to confirm that the query has the things you're requesting.

mcdafydd commented:
I'm going to do a once over on everything again this weekend, but I think it's close enough for re-review.  Thanks!

mcdafydd commented:
I do see one issue that I'll try and address as soon as I can.  The LogToMetric tests and example don't reflect what's required to make that useful:

https://docs.microsoft.com/en-us/azure/azure-monitor/platform/alerts-metric-logs#resource-template-for-metric-alerts-for-logs

To use the LogToMetric API, there needs to be a second resource, an Azure Metric Alert, that alerts on the newly-created metric.  I'll update the tests and example to reflect this.

mcdafydd commented:
Presuming this test passes, I'm ready for another review.

Thanks!

mcdafydd commented:
Oops - I found one additional published limit - 100 resources maximum for a cross-resource query:

https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/azure-subscription-service-limits#log-queries-and-language

mcdafydd commented:
@tombuildsstuff @mbfrahry @katbyte 

Just checking in since this is my first new resource PR.  Is there anything else I should be doing to make sure it is mergeable to master?  I'll be patient, just want to be able to streamline any requirements if possible.

Thanks for all your work on 2.0!

mbfrahry commented on azurerm/internal/services/monitor/tests/data_source_monitor_scheduled_query_rules_alert_test.go:
```diff
@@ -9,7 +9,7 @@ import (
 )
 
-func TestAccDataSourceAzureRMMonitorScheduledQueryRules_alertingAction(t *testing.T) {
+func TestAccDataSourceAzureRMMonitorScheduledQueryRulesAlertingAction(t *testing.T) {
```
Let's keep the `_` that was removed in. It lets us quickly see and grab the prefix when we want to run this subset of tests and has some other minor benefits that warrant keeping it in

mbfrahry commented on website/docs/r/monitor_scheduled_query_rules_alert.html.markdown:
```diff
@@ -25,16 +25,43 @@ resource "azurerm_application_insights" "example" {
+}
+
+# Example: Alerting Action with result count trigger
+# Alert if more than three HTTP requests returned a >= 500 result code
```
I don't think this is going to look quite right on the website. I'd throw a shorter description above the query attribute to explain what's going on

mcdafydd replied:
Let me know if this latest changeset is more aligned with your thoughts.  It's looking a lot nicer to me.

mcdafydd commented on azurerm/helpers/azure/monitor.go:
```diff
@@ -2,8 +2,124 @@ package azure
+	}
+	result = append(result, v)
+
+	return result
```
Thanks!

mcdafydd commented:
@mbfrahry Ready for another peek.

mbfrahry commented:
Thanks for this @mcdafydd! We'll get it into the next release!
